Run integration tests against a live server

rfk opened the pull request:
(not yet ready for merge)

This is an experiment in re-using the integration tests for acceptance testing of a live server instance.  As before, doing the usual `npm test` will run tests against an in-memory Hapi server using inject().  But if you specify TEST_REMOTE environment variable, then the tests will be redirected at a live server.  If you've got it running locally:

```
$ TEST_REMOTE=http://localhost:8080 npm test
```

I'm not thrilled about the naming of functions makeLiveRequest/makeInjectRequest/bindMakeRequest, any suggestions would be appreciated.  But the overall approach seems to work well.

For this to be truly useful, we need to make the tests do proper cleanup so that they can be run multiple times against a single server.  I've made a start on this for the syncstore routes, but auth and blob tests also need to be tweaked.

@zaach @ckarlof thoughts?

rfk committed:
Add a way to delete all the data for a given user.

rfk committed:
Allow running the integration test against a remote server.

If the environment variable TEST_REMOTE is specified, it is interpreted
as the base url of a remote server to be tested.  Integration tests will
send their requests to that server rather than injecting them into a
local Hapi server instance.

rfk committed:
Use randomly-generated auth tokens for testing purposes.

rfk commented:
Adding a stop-gap measure: use randomly-generated auth tokens in the tests.  This ensures each test hits a unique slot in the database so they can be safely run multiple times against the same server.  I think we should still add properly cleanup though.

ckarlof commented:
I like the idea of this. I've always been a little suspicious of (only) inject() style tests, but I appreciate their usefulness.

ckarlof commented:
Hmm "helpers" are usually a bad code smell. How about writing a RequestBuilder?

``` javascript
function RequestBuilder(options) {
  this.server = options.server;
  this.baseUrl = options.baseUrl;
  if (this.server && this.baseUrl) throw some error;
}

RequestBuilder.prototype.makeRequest = function (method, path, options, callback) {
  if (this.baseUrl) {
    this.makeRequestUsingHTTP.apply(this, arguments);
  } else {
    this.makeRequestUsingInject.apply(this, arguments);
  }
}
```

If you don't like the whole [Kingdom of the Nouns](http://steve-yegge.blogspot.com/2006/03/execution-in-kingdom-of-nouns.html) thing, and don't want to instantiate this multiple times, you can just have `baseUrl` and `server` be props on the module and just expose `makeRequest` as an export. Since it's actually making the request and not just building it you could call it `RequestMaker`. Or if you want to keep with your existing structure of actually constructing a `makeRequest` function, you could rewrite the above to use bind again and call it a `MakeRequestBuilder`. FWIW, exposing `bind` in the function name is a leaky abstraction, IMO.

rfk commented:
Thanks @ckarlof, I think I like the RequestBuilder pattern, will try it out next week.  It's essentially the same shape as the code that's there currently, but more explicit and better named.

> FWIW, exposing bind in the function name is a leaky abstraction, IMO

+1, it's icky

zaach commented:
+1 for RequestBuilder

rfk committed:
Refactor bindMakeRequest and friends into a TestClient class.

rfk commented:
Rebased and refactored, this should now be ready for merge.  @zaach r?

After much deliberation on naming, I settled on calling it a "TestClient" object.  I think "testClient.makeRequest()" reads well and captures the intent here.  It has a couple of extra customization hooks, which clean up the code in integration/syncstore.js quite a bit.

I also did away with having each individual test pass in the same server object to the client, since they were all importing it from helpers.js anyway.

zaach commented:
Nice compromise. r+

almet commented on lib/syncstore.js:
```diff
@@ -55,6 +55,11 @@
  *         This method can fail if concurrent writes are attempted.  The error
  *         message in this case will be 'syncstore.writeConflict'.
  *
+ *    deleteUserData(userid, cb(<err>)):
```
Just throwing a comment here to ensure you really wanted to merge this code (allow deletion of user data) here, since it's something different that what the pull request was about (having a live test client).

rfk replied:
Point taken :-)

It was intended as part of the merge, because we need the tests to clean up after themselves properly when running against a live server.  It it would have been better as a separate PR.
